File: bindings/csharp/src/Zag/DiscoverModels.cs
using System.Text.Json.Serialization;

namespace Zag;

/// <summary>Feature support declaration for a provider capability.</summary>
public record FeatureSupport
{
    [JsonPropertyName("supported")]
    public bool Supported { get; init; }

    [JsonPropertyName("native")]
    public bool Native { get; init; }
}

/// <summary>Session log support with completeness level.</summary>
public record SessionLogSupport
{
    [JsonPropertyName("supported")]
    public bool Supported { get; init; }

    [JsonPropertyName("native")]
    public bool Native { get; init; }

    [JsonPropertyName("completeness")]
    public string? Completeness { get; init; }
}

/// <summary>
/// Streaming input support with mid-turn injection semantics.
/// <para>
/// <c>Semantics</c> describes what happens when
/// <c>send_user_message</c> is called while the agent is producing a
/// response on the current turn. One of <c>"queue"</c>,
/// <c>"interrupt"</c>, or <c>"between-turns-only"</c>. <c>null</c> when
/// <c>Supported</c> is <c>false</c>.
/// </para>
/// </summary>
public record StreamingInputSupport
{
    [JsonPropertyName("supported")]
    public bool Supported { get; init; }

    [JsonPropertyName("native")]
    public bool Native { get; init; }

    [JsonPropertyName("semantics")]
    public string? Semantics { get; init; }
}

/// <summary>Size alias mappings (small/medium/large to actual model names).</summary>
public record SizeMappings
{
    [JsonPropertyName("small")]
    public string Small { get; init; } = "";

    [JsonPropertyName("medium")]
    public string Medium { get; init; } = "";

    [JsonPropertyName("large")]
    public string Large { get; init; } = "";
}

/// <summary>All feature flags for a provider.</summary>
public record Features
{
    [JsonPropertyName("interactive")]
    public FeatureSupport Interactive { get; init; } = new();

    [JsonPropertyName("non_interactive")]
    public FeatureSupport NonInteractive { get; init; } = new();

    [JsonPropertyName("resume")]
    public FeatureSupport Resume { get; init; } = new();

    [JsonPropertyName("resume_with_prompt")]
    public FeatureSupport ResumeWithPrompt { get; init; } = new();

    [JsonPropertyName("session_logs")]
    public SessionLogSupport SessionLogs { get; init; } = new();

    [JsonPropertyName("json_output")]
    public FeatureSupport JsonOutput { get; init; } = new();

    [JsonPropertyName("stream_json")]
    public FeatureSupport StreamJson { get; init; } = new();

    [JsonPropertyName("json_schema")]
    public FeatureSupport JsonSchema { get; init; } = new();

    [JsonPropertyName("input_format")]
    public FeatureSupport InputFormat { get; init; } = new();

    [JsonPropertyName("streaming_input")]
    public StreamingInputSupport StreamingInput { get; init; } = new();

    [JsonPropertyName("worktree")]
    public FeatureSupport Worktree { get; init; } = new();

    [JsonPropertyName("sandbox")]
    public FeatureSupport Sandbox { get; init; } = new();

    [JsonPropertyName("system_prompt")]
    public FeatureSupport SystemPrompt { get; init; } = new();

    [JsonPropertyName("auto_approve")]
    public FeatureSupport AutoApprove { get; init; } = new();

    [JsonPropertyName("review")]
    public FeatureSupport Review { get; init; } = new();

    [JsonPropertyName("add_dirs")]
    public FeatureSupport AddDirs { get; init; } = new();

    [JsonPropertyName("max_turns")]
    public FeatureSupport MaxTurns { get; init; } = new();
}

/// <summary>Full capability declaration for a provider.</summary>
public record ProviderCapability
{
    [JsonPropertyName("provider")]
    public string Provider { get; init; } = "";

    [JsonPropertyName("default_model")]
    public string DefaultModel { get; init; } = "";

    [JsonPropertyName("available_models")]
    public List<string> AvailableModels { get; init; } = [];

    [JsonPropertyName("size_mappings")]
    public SizeMappings SizeMappings { get; init; } = new();

    [JsonPropertyName("features")]
    public Features Features { get; init; } = new();
}

/// <summary>Result of resolving a model alias.</summary>
public record ResolvedModel
{
    [JsonPropertyName("input")]
    public string Input { get; init; } = "";

    [JsonPropertyName("resolved")]
    public string Resolved { get; init; } = "";

    [JsonPropertyName("is_alias")]
    public bool IsAlias { get; init; }

    [JsonPropertyName("provider")]
    public string Provider { get; init; } = "";
}

File: bindings/kotlin/src/main/kotlin/zag/DiscoverModels.kt
package zag

import kotlinx.serialization.*

/**
 * Feature support declaration for a provider capability.
 */
@Serializable
data class FeatureSupport(
    val supported: Boolean = false,
    @SerialName("native") val isNative: Boolean = false,
)

/**
 * Session log support with completeness level.
 */
@Serializable
data class SessionLogSupport(
    val supported: Boolean = false,
    @SerialName("native") val isNative: Boolean = false,
    val completeness: String? = null,
)

/**
 * Size alias mappings (small/medium/large to actual model names).
 */
@Serializable
data class SizeMappings(
    val small: String = "",
    val medium: String = "",
    val large: String = "",
)

/**
 * All feature flags for a provider.
 */
@Serializable
data class Features(
    val interactive: FeatureSupport = FeatureSupport(),
    @SerialName("non_interactive") val nonInteractive: FeatureSupport = FeatureSupport(),
    val resume: FeatureSupport = FeatureSupport(),
    @SerialName("resume_with_prompt") val resumeWithPrompt: FeatureSupport = FeatureSupport(),
    @SerialName("session_logs") val sessionLogs: SessionLogSupport = SessionLogSupport(),
    @SerialName("json_output") val jsonOutput: FeatureSupport = FeatureSupport(),
    @SerialName("stream_json") val streamJson: FeatureSupport = FeatureSupport(),
    @SerialName("json_schema") val jsonSchema: FeatureSupport = FeatureSupport(),
    @SerialName("input_format") val inputFormat: FeatureSupport = FeatureSupport(),
    @SerialName("streaming_input") val streamingInput: FeatureSupport = FeatureSupport(),
    val worktree: FeatureSupport = FeatureSupport(),
    val sandbox: FeatureSupport = FeatureSupport(),
    @SerialName("system_prompt") val systemPrompt: FeatureSupport = FeatureSupport(),
    @SerialName("auto_approve") val autoApprove: FeatureSupport = FeatureSupport(),
    val review: FeatureSupport = FeatureSupport(),
    @SerialName("add_dirs") val addDirs: FeatureSupport = FeatureSupport(),
    @SerialName("max_turns") val maxTurns: FeatureSupport = FeatureSupport(),
)

/**
 * Full capability declaration for a provider.
 */
@Serializable
data class ProviderCapability(
    val provider: String = "",
    @SerialName("default_model") val defaultModel: String = "",
    @SerialName("available_models") val availableModels: List<String> = emptyList(),
    @SerialName("size_mappings") val sizeMappings: SizeMappings = SizeMappings(),
    val features: Features = Features(),
)

/**
 * Result of resolving a model alias.
 */
@Serializable
data class ResolvedModel(
    val input: String = "",
    val resolved: String = "",
    @SerialName("is_alias") val isAlias: Boolean = false,
    val provider: String = "",
)
